Changes to rebel attack

wurzel0701 opened the pull request:
## What type of PR is this.
1. [X] Bug
2. [X] Change
3. [X] Enhancement

### What have you changed and why?
Information:
Multiple smaller changes to rebelAttack and functions around that
- Removed that block that Invaders cannot attack at war level 1 (see dev meeting at 04-07-2020)
- The target side is now choosen based on aggression and captured location (see dev meeting at 04-07-2020)
- Aggression for AI vs AI is now inverted to have the factions attack more aggressive (see dev meeting at 04-07-2020)
- Added some blocks for aggro calculation of not existing factions
- Added autoresolving for small and large attacks, when no player is near and the target is not player hold
- Remove a double call of timing CA, so attacks are 60 minutes apart on average (reduced from 120)

### Please specify which Issue this PR Resolves.
No issues opened for that

### Please verify the following and ensure all checks are completed.

1. [X] Have you loaded the mission in singleplayer?
2. [X] Have you loaded the mission in LAN host?
3. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [ ] No
2. [X] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps: 
Trigger the rebelAttack function by executing ```[Invaders] spawn A3A_fnc_rebelAttack``` (or replace Invaders with Occupants) or wait till the aggression loop triggers it automatically.
********************************************************
Notes: You may want to wait till the support rework drops, as the original attacks are horrible and might use even more resources than the old once. However, mosts attacks are outside of the players vision and get autoresolved.

Havent checked the balance yet, the added amount of units might be a bit high

CalebSerafin commented on A3-Antistasi/functions/Base/fn_aggressionUpdateLoop.sqf:
```diff
@@ -43,24 +43,27 @@ while {true} do
-        [3600, Occupants] call A3A_fnc_timingCA;
-        if (!bigAttackInProgress) then
+        //Update attack countdown for occupants and execute attack if needed
+        attackCountdownOccupants = attackCountdownOccupants - (60 * (0.5 + (aggressionOccupants/100)));
```
This requires the use of an accurate delay above:

This line here: https://github.com/official-antistasi-community/A3-Antistasi/blob/73cb64f19b82f5e1bed31f28701b3a415ad962ec/A3-Antistasi/functions/Base/fn_aggressionUpdateLoop.sqf#L16
should be `uiSleep 60` as it seems to be an important timing component. `sleep`'s duration  can vary a lot depending on performance.

wurzel0701 replied:
I dont think this is needed, even if it is off by 10 seconds which I doubt, its no problem, no need to make it more precise

jaj22 commented:
Nothing very line-specific so I'm gonna go long-form.

**Timing**
I feel like you removed the better timingCA call here (although it needed adjustment for the simulation). As far as I can tell, the attackCountdownXXX values are still decremented during a major attack, while 60 minutes are added if the countdown hits zero with an attack in progress, so if you spend 55 minutes fending off an attack, you get a break of five minutes before the next one, while if you spend 61 minutes then you get a 59 minute break.

Additionally, once you get to the lategame, it looks like you're potentially getting hit by both enemy factions at half-hour intervals, even if you clear the attacks rapidly. They'll be attacking mostly with garbage vehicles, but it sounds like a massive annoyance for a small group. I might be missing something here.

Also I think you missed the similar timingCA call in invaderPunish.

**Target side selection**
Seems reasonable, although an oddity is that towns are counted towards the total, and invaders can't hold towns. This may result in occupants attacking the players more than intended, and in some cases an optimal strategy that involves blowing up your own radio towers.

**Inverse aggro for enemy vs enemy**
Something that looked reasonable on the surface but makes the enemies do weird things in practice. An oversized or undersized attack is still oversized or undersized regardless of what's happening on the other side of the map. Would recommend a fixed aggro value for enemy vs enemy instead. Using aggro in target side selection is a sufficient adjustment.

**Easy attacks**
I really don't like these in general. The trigger point is very arbitrary (keep three easy targets and you're good), they complicate the code (see city attacks later), and have a massive potential performance hit on the rare occasion that they do happen (4*6 vehicle patrolCAs spawning simultaneously, mmm). With this simulation system they won't happen anyway with AI vs AI because the garrisons get filled after a capture. I think the better approach is to have a smaller timingCA bump if a QRF is suitable for the (single) chosen target.

**City attacks**
Not new, but I think I figured out what happens with them. Cities are never considered to be easy targets, so they take priority over easy targets, even when the easy targets are considered more worthwhile by the metric. Also cities tend to have low difficulty values in general, as they have no garrisons. The algorithm is unconcerned that attacking a city is a waste of time for the occupants.

**Vehicle loss**
Real attacks often result in the destruction of several attacking helicopters, and occasionally other vehicles. Simulated attacks do not, so enemy factions will accumulate more vehicles for use against the rebels. Maybe less of an issue than economicsAI having almost no player-count adjustment, but it should be considered.

Real AI vs AI attacks do sometimes lose outright too, not just in the timeout cases. Some fairly common cases include a single AA tank murdering an entire air attack or attacks failing because they lack effective vehicles. QRF mini-attacks almost universally failed with AI vs AI before I blocked them. I don't think auto-win attacks are unreasonable because attacks *should* be sized to win, but equally if a factions lacks sufficient vehicles then it shouldn't attack, simulated or otherwise.

**Garrison filling**
This is reasonable. Main issue is that it's missing from the unsimulated attacks, probably because you'd need to filter all the attacking infantry so that garrisons don't end up with weird units. Large garrisons take an age to fill up afterwards, which makes player recaptures very easy as long as they despawn the target area first. I guess that can be fixed later.

wurzel0701 commented on A3-Antistasi/functions/Base/fn_rebelAttack.sqf:
```diff
@@ -14,66 +14,79 @@
+    private _allTargetsCount = _playersHold + _ownHold + _enemyAIHold;
+
+    _playersHold = _playersHold / _allTargetsCount;
+    _enemyAIHold = _enemyAIHold / _allTargetsCount;
```
Changed

wurzel0701 commented:
**Timing**
Fixed the timingCA calls based on our discussion on github. The timingCA call happens after the large attack finished and instantly if the fight is simulated. For the four small attacks it gets called instantly, but with 90 minutes instead of 60, so even if the small attack times out, it has a 30 minute pause till the next attack.

If a large attack is blocked, it now adds a 10 minute delay instead of the full 60 minutes, to wait only a bit until the other faction can attack. For example CSAT wants to start an attack, but NATO is currently doing some, CSAT will check every ten minutes if NATO finished and start their own after that.

**Target side selection**
Get cities out of the calculation, it is now purely dependent on the military locations.

**Inverse aggro for enemy vs enemy**
I would prefer to keep it like this, these over/underpowered attacks only happen in range of players and lead to some nice gameplay, as you can step into a fight where a small attacks weakens a target for you or an attack where one side goes full force on the other and you can watch or retreat to save you. Had a few funny moments with these overpowered attacks, while underpowered attacks are quite rare as aggro doesnt stays up for too long.

**Easy attacks**
Legacy code, I kept it in as punishment for players who dont garrisons their locations. When they spawn near players they are three minutes apart to keep the performance impact as low as possible. 

**City attacks**
City attacks are now less frequent as I nerfed their modifier (airports with same garrison can now be 40 times further away until the code thinks the city is better). However Invader attacks are more powerful anyways as they destroy the city, Occupants attacks now punch down the support in the whole area, reverting work of multiple hours if not fought back

**Vehicle loss**
The AI runs out of vehicles quick enough already, I dont think we should make it even worse unless the game gets unplayable by it.

**Garrison filling**
For real attacks, the units gets added twice already (maybe even overwriting itself) but thats not part of this PR. But you are right the reinforcement rates are shitty, especially on larger maps

CalebSerafin commented on A3-Antistasi/functions/CREATE/fn_wavedCA.sqf:
```diff
@@ -717,15 +717,49 @@ while {(_waves > 0)} do
+                    {
+                        private _distance = (getMarkerPos _mrkDestination) distance2D (getMarkerPos _x);
+                        private _supportChange = [0, 0];
+                        if(_distance < 2500) then
```
Ln728-Ln747 (Can't multi line select on mobile)
Looks like the if branching can be replaced by an math equation with floor to keep round numbers.

CalebSerafin commented on A3-Antistasi/functions/CREATE/fn_wavedCA.sqf:
```diff
@@ -717,15 +717,49 @@ while {(_waves > 0)} do
 				_mrkD setMarkerColor colorOccupants;
 				garrison setVariable [_mrkDestination,[],true];
-				};
+			};
```
Check Indentation with surrounding context. In github, it looks as if it lines up with the proceeding brace. This indicates mixing of tabs with spaces, or sourunding indentation is at fault.

jaj22 reviewed (changes requested):
Most of my complaints are with the previous rebelAttack PR, and almost everything in this one is an improvement, so not much reason to avoid pushing it into 2.3.1. I would recommend flipping those wavedCA calls back to server spawns though.

Bob-Murphy commented:
Changed the branch to 2.3.1

wurzel0701 commented on A3-Antistasi/functions/Base/fn_rebelAttack.sqf:
```diff
@@ -389,24 +407,45 @@ else
+        {
+            //Sending real attack, execute the fight
+            [2, format ["Starting waved attack with %1 waves from %2 to %3", _waves, _attackOrigin, _attackTarget], _fileName] call A3A_fnc_log;
+            [[_attackTarget, _attackOrigin, _waves],"A3A_fnc_wavedCA"] remoteExec ["A3A_fnc_scheduler",2];
```
Reverted back to server execution

wurzel0701 commented on A3-Antistasi/functions/Base/fn_rebelAttack.sqf:
```diff
@@ -299,11 +283,45 @@ if(count _easyTargets >= 4) then
+            {
+                params ["_side", "_target"];
+                sleep 10;
+                private _squads = 3 + round (random 2);
```
Reduced by 1

wurzel0701 commented on A3-Antistasi/functions/Base/fn_rebelAttack.sqf:
```diff
@@ -389,24 +407,45 @@ else
+            {
+                params ["_side", "_target"];
+                sleep 10;
+                private _squads = 5 + round (random 3);
```
Reduced by 1

wurzel0701 commented on A3-Antistasi/functions/Base/fn_rebelAttack.sqf:
```diff
@@ -389,24 +407,45 @@ else
-        //Why not using the scheduler here?
-        [_attackTarget, _attackOrigin] spawn A3A_fnc_invaderPunish;
+        [2, format ["Starting punishment mission from %1 to %2", _attackOrigin, _attackTarget], _fileName] call A3A_fnc_log;
+        [[_attackTarget, _attackOrigin], "A3A_fnc_invaderPunish"] remoteExec ["A3A_fnc_scheduler", 2];
```
Reverted to server execution

wurzel0701 commented on A3-Antistasi/functions/Base/fn_rebelAttack.sqf:
```diff
@@ -196,17 +183,16 @@ if (count _availableTargets == 0) exitWith
+        case (_target in factories): {_targetMultiplier = 0.5};
         case (_target in seaports): {_targetMultiplier = 0.7};
-        case (_target in citiesX): {_targetMultiplier = 0.9};
+        case (_target in citiesX): {_targetMultiplier = 2};
```
Yeah, garrisoned cities are relatively safe (unless they are really close to the enemy). But most players dont do that.

Anyways, with the increase support penalty it should be a one time attack

wurzel0701 commented on A3-Antistasi/functions/CREATE/fn_wavedCA.sqf:
```diff
@@ -717,15 +717,49 @@ while {(_waves > 0)} do
+                        };
+                    };
+                } forEach citiesX;
+				[100,-100,_mrkDestination] remoteExec ["A3A_fnc_citySupportChange",2];
```
It might be brutal, but cities can be defended quite easily with a few units. 

If it doesnt punches their support up, one radio tower can revert all effects of the attack in one tick. The attack punches the city down to 0/0 and the radio tower sets it to 2/0 and flips it back to rebel side, having repeating attacks over and over again

jaj22 replied:
I'd be a lot happier with the idea of garrisoning cities if it stopped them getting flipped by enemy radio towers. Garrisoning cities isn't currently a sensible choice unless you control the radio tower or you can maintain enough activity near the town to prevent it flipping, as otherwise you'll likely lose more HR from vanished garrisons than you gain from the city.

Chance of getting attacked in the early game should be low anyway if you're playing reb vs gov vs inv, which there's no reason to avoid after this patch. So it'll do for now.

wurzel0701 replied:
Sounds like a neat idea. With the new system we would also be able to tie it into the propaganda value and affect support directly (see data-dump channel)

Will also add it to meeting points

wurzel0701 commented:
Changed everything related to the comments beside style related stuff (Atom refuses to show me the differences)

According to the meeting point from 15-08-2020

4) Auto win based on percentage - higher chance for attacker - calculation and substraction of lost units - after the attack fill the garrison

These are the things missing right now, but I dont like these changes and will not do it. I will leave some comments how I would approach it. 

Autowin based on percentage: Before executing the marker flip check against a chance and dump the calculation if false. This will drop alot of calculation and will not gain anything for the player as nothing happens (Alot of CPU usage for nothing)

Higher chance for attacker: Unclear, I think it means Invader, but you know what to do then

Calculation and substraction of lost units: Good luck with that

After attack fill garrison: Based on the point above just add the remaining units as I added the current fixed units

This adds the apperantly long requested feature of simulating most attacks. If it is really this important @Bob-Murphy will find someone to do this.

jaj22 commented:
- Higher chance for attacker: Unclear, I think it means Invader, but you know what to do then

It's important that Invaders take ground, on average. I think that's generally true here due to map layout. Rebels will typically engage occupants in the early to mid game, and because the chance of a side attacking rebels scales faster with aggro than the rate of attacking, invaders will attack occupants more rapidly than vice versa, and so will gain ground (attacks now being auto-win, so only the rate matters).

There may be a problem if rebels keep their aggro and territory control near zero for an extended period. In this case, occupants and invaders are now equals, but the players can swing the attack rate in favour of occupants by failing missions (all of which will be occupant-sided), causing invaders to get pushed back into the ocean.

It's not necessarily much of a problem, given that invaders should start taking territory (autowin even against airfields) as soon as rebels start playing CTI, but there's an open question about what we want the early game to look like.

I think it'll do for the moment anyway. Curious what happens in practice.

Bob-Murphy commented:
I will merge this one dispite the small changes not done.
Having the autoresolving has priority.
If we find that the adjustments are still needed/wanted in the future, we can introduce them later.

> These are the things missing right now, but I dont like these changes and will not do it. I will leave some comments how I would approach it.

> Autowin based on percentage: Before executing the marker flip check against a chance and dump the calculation if false. This will drop alot of calculation and will not gain anything for the player as nothing happens (Alot of CPU usage for nothing)

> Higher chance for attacker: Unclear, I think it means Invader, but you know what to do then

> Calculation and substraction of lost units: Good luck with that

> After attack fill garrison: Based on the point above just add the remaining units as I added the current fixed units

> This adds the apperantly long requested feature of simulating most attacks. If it is really this important @Bob-Murphy will find someone to do this.
